File: packages/cucumber-ruby/lib/kensho/_schema.rb
# frozen_string_literal: true

# Vendored slice of the Kensho v1 schema contract.
#
# We deliberately do not depend on the JS @kaizenreport/kensho-schema package.
# Adapters need to stay tiny and standalone, so we re-implement the few things
# the formatter cares about: the stable case-id hash, the status enum, the
# attachment kind/MIME tables, and the environment-capture function.

require 'rbconfig'

module Kensho
  module Schema
    SCHEMA_VERSION = 'kensho/v1'

    STATUS = %w[pass fail broken skip].freeze
    STEP_STATUS = %w[pass fail skip].freeze
    SEVERITY = %w[blocker critical normal minor trivial].freeze
    ATTACHMENT_KINDS = %w[
      screenshot video trace har text json html dom-snapshot log
    ].freeze

    MIME_BY_EXT = {
      '.png'  => 'image/png',
      '.jpg'  => 'image/jpeg',
      '.jpeg' => 'image/jpeg',
      '.webp' => 'image/webp',
      '.webm' => 'video/webm',
      '.mp4'  => 'video/mp4',
      '.zip'  => 'application/zip',
      '.html' => 'text/html',
      '.json' => 'application/json',
      '.txt'  => 'text/plain',
      '.log'  => 'text/plain',
      '.har'  => 'application/json'
    }.freeze

    KIND_BY_EXT = {
      '.png'  => 'screenshot',
      '.jpg'  => 'screenshot',
      '.jpeg' => 'screenshot',
      '.webp' => 'screenshot',
      '.webm' => 'video',
      '.mp4'  => 'video',
      '.zip'  => 'trace',
      '.html' => 'html',
      '.json' => 'json',
      '.txt'  => 'text',
      '.log'  => 'log',
      '.har'  => 'har'
    }.freeze

    FNV_OFFSET_1 = 0x811c9dc5
    FNV_OFFSET_2 = 0x01000193
    FNV_PRIME_1  = 0x01000193
    FNV_PRIME_2  = 0x85ebca6b
    MASK32       = 0xffffffff

    # Mirrors stableCaseId from packages/schema/index.js byte-for-byte:
    # double FNV-1a with two different secondary primes so the two 32-bit
    # chunks come from independent rolling states. Must stay byte-compatible
    # with the JS implementation or test history won't line up across
    # adapters.
    def self.stable_case_id(full_name, file_path)
      s = "#{full_name || ''}::#{file_path || ''}"
      h1 = FNV_OFFSET_1
      h2 = FNV_OFFSET_2
      s.each_codepoint do |c|
        h1 = ((h1 ^ c) * FNV_PRIME_1) & MASK32
        h2 = ((h2 ^ c) * FNV_PRIME_2) & MASK32
      end
      format('tc_%08x%08x', h1, h2)
    end

    def self.kind_and_mime_for(path)
      ext = File.extname(path.to_s).downcase
      [KIND_BY_EXT[ext] || 'text', MIME_BY_EXT[ext] || 'application/octet-stream']
    end

    def self.normalize_os
      host = RbConfig::CONFIG['host_os'].to_s.downcase
      return 'linux'  if host.include?('linux')
      return 'darwin' if host.include?('darwin')
      return 'win32'  if host.include?('mswin') || host.include?('mingw') || host.include?('cygwin')

      host
    end

    # Normalize SSH-style git URLs to https. ``git@github.com:foo/bar.git``
    # becomes ``https://github.com/foo/bar``; trailing ``.git`` is dropped.
    def self.normalize_git_url(u)
      return nil if u.nil? || u.empty?

      m = u.match(%r{^(?:ssh://)?git@([^:/]+)[:/](.+?)(?:\.git)?$})
      return "https://#{m[1]}/#{m[2]}" if m

      u.sub(/\.git\z/, '')
    end

    # CI / environment metadata for run.env. Matches the helpers in the JS
    # adapters so a Kensho report looks the same regardless of language.
    def self.env_info(framework_version: nil)
      ci_env = ENV['CI']
      ci =
        if ci_env && ENV['GITHUB_ACTIONS']
          'github-actions'
        elsif ci_env && ENV['CIRCLECI']
          'circleci'
        elsif ci_env && ENV['GITLAB_CI']
          'gitlab'
        elsif ci_env && ENV['JENKINS_URL']
          'jenkins'
        elsif ci_env && ENV['BUILDKITE']
          'buildkite'
        elsif ci_env && ENV['TF_BUILD']
          'azure-devops'
        elsif ci_env
          'unknown'
        else
          'local'
        end

      branch = ENV['GITHUB_REF_NAME'] || ENV['CIRCLE_BRANCH'] ||
               ENV['CI_COMMIT_REF_NAME'] || ENV['BUILDKITE_BRANCH']
      commit = ENV['GITHUB_SHA'] || ENV['CIRCLE_SHA1'] ||
               ENV['CI_COMMIT_SHA'] || ENV['BUILDKITE_COMMIT']
      author = ENV['KR_AUTHOR'] || ENV['GITHUB_ACTOR']
      commit_msg = ENV['KR_COMMIT_MSG']

      run_url = ENV['KR_RUN_URL'] ||
        if ENV['GITHUB_SERVER_URL'] && ENV['GITHUB_REPOSITORY'] && ENV['GITHUB_RUN_ID']
          "#{ENV['GITHUB_SERVER_URL']}/#{ENV['GITHUB_REPOSITORY']}/actions/runs/#{ENV['GITHUB_RUN_ID']}"
        elsif ENV['CIRCLE_BUILD_URL']
          ENV['CIRCLE_BUILD_URL']
        elsif ENV['CI_JOB_URL']
          ENV['CI_JOB_URL']
        elsif ENV['BUILD_URL']
          ENV['BUILD_URL']
        elsif ENV['BUILDKITE_BUILD_URL']
          ENV['BUILDKITE_BUILD_URL']
        end

      # repoUrl — KR_REPO_URL override → GitHub Actions / GitLab / Bitbucket /
      # Azure / SSH-style URLs from CircleCI / Buildkite / Jenkins (normalized).
      gh_server = ENV['GITHUB_SERVER_URL']
      gh_repo   = ENV['GITHUB_REPOSITORY']
      repo_url = ENV['KR_REPO_URL'] ||
                 (gh_server && gh_repo ? "#{gh_server}/#{gh_repo}" : nil) ||
                 ENV['CI_PROJECT_URL'] ||
                 ENV['BITBUCKET_GIT_HTTP_ORIGIN'] ||
                 normalize_git_url(ENV['BUILD_REPOSITORY_URI']) ||
                 normalize_git_url(ENV['CIRCLE_REPOSITORY_URL'] ||
                                   ENV['BUILDKITE_REPO'] ||
                                   ENV['GIT_URL'])

      info = {
        'ci' => ci,
        'os' => normalize_os,
        'arch' => RbConfig::CONFIG['host_cpu'].to_s
      }

      ruby_version = "#{RUBY_VERSION}p#{defined?(RUBY_PATCHLEVEL) ? RUBY_PATCHLEVEL : '0'}"
      info['vars'] = { 'rubyVersion' => ruby_version }
      info['vars']['frameworkVersion'] = framework_version if framework_version

      info['branch']     = branch     if branch
      info['commit']     = commit     if commit
      info['commitMsg']  = commit_msg if commit_msg
      info['author']     = author     if author
      info['runUrl']     = run_url    if run_url
      info['repoUrl']    = repo_url   if repo_url
      os_version = RbConfig::CONFIG['host_os']
      info['osVersion'] = os_version if os_version

      [
        ['KR_STAGE', 'stage'],
        ['KR_BASE_URL', 'baseUrl'],
        ['KR_APP_VERSION', 'appVersion'],
        ['KR_BUILD_NUMBER', 'buildNumber'],
        ['KR_RELEASE', 'release'],
        ['KR_REGION', 'region'],
        ['KR_LOCALE', 'locale'],
        ['KR_TRIGGER', 'trigger'],
        ['KR_FEATURE', 'feature']
      ].each do |env_var, key|
        v = ENV[env_var]
        info[key] = v if v && !v.empty?
      end

      info
    end

    def self.iso_now
      Time.now.utc.strftime('%Y-%m-%dT%H:%M:%S.%LZ')
    end

    def self.iso_from_seconds(secs)
      Time.at(secs).utc.strftime('%Y-%m-%dT%H:%M:%S.%LZ')
    end

    def self.slugify(name)
      s = name.to_s.downcase.strip
      s = s.gsub(/[^a-z0-9_-]+/, '-').gsub(/\A-+|-+\z/, '')
      s.empty? ? 'unknown' : s
    end

    def self.default_run_id
      "run_#{Time.now.utc.strftime('%Y%m%d%H%M%S')}"
    end
  end
end

File: packages/robot/src/kensho_robot/_schema.py
"""Vendored slice of the Kensho v1 schema contract.

We deliberately do not depend on the JS ``@kaizenreport/kensho-schema`` package.
Adapters need to stay tiny and standalone, so we re-implement the few things
the listener cares about: the stable case-id hash, the status enum, the
attachment kind/MIME tables, and the environment-capture function.

This file is a verbatim copy of ``kensho_pytest._schema`` so the FNV-1a-based
``stable_case_id`` matches byte-for-byte across adapters.
"""

from __future__ import annotations

import os
import platform
import sys
from pathlib import Path
from typing import Dict, Optional

SCHEMA_VERSION = "kensho/v1"

STATUS = ("pass", "fail", "broken", "skip")
STEP_STATUS = ("pass", "fail", "skip")
SEVERITY = ("blocker", "critical", "normal", "minor", "trivial")
ATTACHMENT_KINDS = (
    "screenshot",
    "video",
    "trace",
    "har",
    "text",
    "json",
    "html",
    "dom-snapshot",
    "log",
)

MIME_BY_EXT: Dict[str, str] = {
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".webp": "image/webp",
    ".webm": "video/webm",
    ".mp4": "video/mp4",
    ".zip": "application/zip",
    ".html": "text/html",
    ".json": "application/json",
    ".txt": "text/plain",
    ".log": "text/plain",
    ".har": "application/json",
}

KIND_BY_EXT: Dict[str, str] = {
    ".png": "screenshot",
    ".jpg": "screenshot",
    ".jpeg": "screenshot",
    ".webp": "screenshot",
    ".webm": "video",
    ".mp4": "video",
    ".zip": "trace",
    ".html": "html",
    ".json": "json",
    ".txt": "text",
    ".log": "log",
    ".har": "har",
}

_FNV_OFFSET_1 = 0x811C9DC5
_FNV_OFFSET_2 = 0x01000193
_PRIME_1 = 0x01000193
_PRIME_2 = 0x85EBCA6B
_MASK = 0xFFFFFFFF


def stable_case_id(full_name: str, file_path: Optional[str]) -> str:
    """Mirror ``stableCaseId`` from ``packages/schema/index.js``.

    Double FNV-1a so two 32-bit chunks come from independent rolling states.
    Must stay byte-for-byte identical to the JS implementation; that's how
    the platform correlates Robot, pytest, Playwright, and Go runs of the
    same suite to a single test history.
    """
    s = (full_name or "") + "::" + (file_path or "")
    h1 = _FNV_OFFSET_1
    h2 = _FNV_OFFSET_2
    for ch in s:
        c = ord(ch)
        h1 = ((h1 ^ c) * _PRIME_1) & _MASK
        h2 = ((h2 ^ c) * _PRIME_2) & _MASK
    return "tc_" + format(h1, "08x") + format(h2, "08x")


def env_info() -> Dict[str, object]:
    """Collect CI / environment metadata for ``run.env``."""
    is_ci = bool(os.environ.get("CI"))
    if is_ci and os.environ.get("GITHUB_ACTIONS"):
        ci = "github-actions"
    elif is_ci and os.environ.get("CIRCLECI"):
        ci = "circleci"
    elif is_ci and os.environ.get("GITLAB_CI"):
        ci = "gitlab"
    elif is_ci and os.environ.get("JENKINS_URL"):
        ci = "jenkins"
    elif is_ci and os.environ.get("BUILDKITE"):
        ci = "buildkite"
    elif is_ci and os.environ.get("TF_BUILD"):
        ci = "azure-devops"
    elif is_ci:
        ci = "unknown"
    else:
        ci = "local"

    branch = (
        os.environ.get("GITHUB_REF_NAME")
        or os.environ.get("CIRCLE_BRANCH")
        or os.environ.get("CI_COMMIT_REF_NAME")
        or os.environ.get("BUILDKITE_BRANCH")
    )
    commit = (
        os.environ.get("GITHUB_SHA")
        or os.environ.get("CIRCLE_SHA1")
        or os.environ.get("CI_COMMIT_SHA")
        or os.environ.get("BUILDKITE_COMMIT")
    )
    author = os.environ.get("KR_AUTHOR") or os.environ.get("GITHUB_ACTOR")
    commit_msg = os.environ.get("KR_COMMIT_MSG")

    run_url: Optional[str] = os.environ.get("KR_RUN_URL")
    gh_server = os.environ.get("GITHUB_SERVER_URL")
    gh_repo = os.environ.get("GITHUB_REPOSITORY")
    gh_run_id = os.environ.get("GITHUB_RUN_ID")
    if not run_url:
        if gh_server and gh_repo and gh_run_id:
            run_url = f"{gh_server}/{gh_repo}/actions/runs/{gh_run_id}"
        elif os.environ.get("CIRCLE_BUILD_URL"):
            run_url = os.environ["CIRCLE_BUILD_URL"]
        elif os.environ.get("CI_PIPELINE_URL"):
            run_url = os.environ["CI_PIPELINE_URL"]
        elif os.environ.get("CI_JOB_URL"):
            run_url = os.environ["CI_JOB_URL"]
        elif os.environ.get("BUILD_URL"):
            run_url = os.environ["BUILD_URL"]
        elif os.environ.get("BUILDKITE_BUILD_URL"):
            run_url = os.environ["BUILDKITE_BUILD_URL"]

    # repoUrl — KR_REPO_URL override → GitHub Actions / GitLab / Bitbucket /
    # Azure / SSH-style URLs from CircleCI / Buildkite / Jenkins (normalized).
    def _normalize_git_url(u: Optional[str]) -> Optional[str]:
        if not u:
            return None
        import re
        m = re.match(r"^(?:ssh://)?git@([^:/]+)[:/](.+?)(?:\.git)?$", u)
        if m:
            return f"https://{m.group(1)}/{m.group(2)}"
        return u[:-4] if u.endswith(".git") else u

    repo_url: Optional[str] = os.environ.get("KR_REPO_URL")
    if not repo_url:
        if gh_server and gh_repo:
            repo_url = f"{gh_server}/{gh_repo}"
        elif os.environ.get("CI_PROJECT_URL"):
            repo_url = os.environ["CI_PROJECT_URL"]
        elif os.environ.get("BITBUCKET_GIT_HTTP_ORIGIN"):
            repo_url = os.environ["BITBUCKET_GIT_HTTP_ORIGIN"]
        elif os.environ.get("BUILD_REPOSITORY_URI"):
            repo_url = _normalize_git_url(os.environ["BUILD_REPOSITORY_URI"])
        else:
            repo_url = _normalize_git_url(
                os.environ.get("CIRCLE_REPOSITORY_URL")
                or os.environ.get("BUILDKITE_REPO")
                or os.environ.get("GIT_URL")
            )

    info: Dict[str, object] = {
        "ci": ci,
        "os": _normalize_os(),
        "arch": platform.machine() or sys.platform,
        "pythonVersion": platform.python_version(),
    }
    if branch:
        info["branch"] = branch
    if commit:
        info["commit"] = commit
    if commit_msg:
        info["commitMsg"] = commit_msg
    if author:
        info["author"] = author
    if run_url:
        info["runUrl"] = run_url
    if repo_url:
        info["repoUrl"] = repo_url
    os_version = platform.release()
    if os_version:
        info["osVersion"] = os_version

    for var, key in (
        ("KR_STAGE", "stage"),
        ("KR_BASE_URL", "baseUrl"),
        ("KR_APP_VERSION", "appVersion"),
        ("KR_BUILD_NUMBER", "buildNumber"),
        ("KR_RELEASE", "release"),
        ("KR_REGION", "region"),
        ("KR_LOCALE", "locale"),
        ("KR_TRIGGER", "trigger"),
        ("KR_FEATURE", "feature"),
    ):
        v = os.environ.get(var)
        if v:
            info[key] = v
    return info


def _normalize_os() -> str:
    p = sys.platform
    if p.startswith("linux"):
        return "linux"
    if p == "darwin":
        return "darwin"
    if p in ("win32", "cygwin"):
        return "win32"
    return p


def kind_and_mime_for(path: str) -> tuple:
    """Return ``(kind, mimeType)`` defaulting to ``('text', 'application/octet-stream')``."""
    ext = Path(path).suffix.lower()
    kind = KIND_BY_EXT.get(ext, "text")
    mime = MIME_BY_EXT.get(ext, "application/octet-stream")
    return kind, mime
